fdm-app/app/root.tsx

```diff
@@ -77,5 +81,58 @@ export function Layout() {
 }
 
 export default function App() {
-    return <Outlet />
+    return <Layout />
+}
+
+export function ErrorBoundary({ error }: Route.ErrorBoundaryProps) {
+    const location = useLocation()
+    const page = location.pathname
+    const timestamp = new Date().toISOString()
+
+    if (isRouteErrorResponse(error)) {
+        const clientErrors = [400, 401, 403, 404]
+        if (clientErrors.includes(error.status)) {
+            return (
+                <ErrorBlock
+                    status={404} // Show 404 in case user is not authorized to access page
+                    message={error.statusText}
+                    stacktrace={error.data}
+                    page={page}
+                    timestamp={timestamp}
+                />
+            )
+        }
+        return (
+            <ErrorBlock
+                status={error.status}
+                message={error.statusText}
+                stacktrace={error.data}
+                page={page}
+                timestamp={timestamp}
+            />
+        )
+    }
+    if (error instanceof Error) {
+        return (
+            <ErrorBlock
+                status={500}
+                message={error.message}
+                stacktrace={error.stack}
+                page={page}
+                timestamp={timestamp}
+            />
+        )
+    }
+    if (error === null) {
+        return null
```

Comment:
errorblock is not automatically handlign null cases correctly? Its just a few lines, so not a problem.